### PR TITLE
feat: setup proxy lib foundation

### PR DESCRIPTION
**Pull request by xilosada**
Setup proxy lib structure for testing
In this initial commit I just added the foundation of the NEAR lib. It implements a basic smartcontract to request the member list from the config contract. Tests and code will change a lot in the following PRs. 
NOTE:
This PR also downgrade near-workspaces version to support darwin x86-64

### CODE REVIEW

**Comment by xilosada** [comment]
@miraclx I'll replace the errors with eyre in the next commit

File: crates/context/config/src/repr.rs
Comment (miraclx) [inline crates/context/config/src/repr.rs]: I think deriving `Hash` for `Repr<T>` is best so it defers to `T: Hash` ..._

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: _... although it looks like `Repr<T>: Hash` is only needed because of this.

`Repr` is only necessary for wire serialization (JSON). For state serialization `SignerId` already implements `Borsh{Se,Dese}rialize`.

```suggestion
    pub approvals: IterableMap<ProposalId, HashSet<SignerId>>,
```

File: contracts/context-config/src/query.rs
Comment (danwilliams) [inline contracts/context-config/src/query.rs:51]: We may want to think about returning a `Result` here.

File: contracts/proxy-lib/src/ext_config.rs
Comment (danwilliams) [inline contracts/proxy-lib/src/ext_config.rs]: Should this be `ConfigContract`?

File: contracts/proxy-lib/src/lib.rs
Comment (danwilliams) [inline contracts/proxy-lib/src/lib.rs]: Ideally we would turn this into Rustdoc documentation, i.e. using `///` and then following the standard formatting approach for function documentation 🙂

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: the `#[init]` annotation expands to this:

```rust
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn init() {
    ::near_sdk::env::setup_panic_hook();
    if ::near_sdk::env::attached_deposit().as_yoctonear() != 0 {
        ::near_sdk::env::panic_str("Method init doesn't accept deposit");
    }
    if ::near_sdk::env::state_exists() {
        ::near_sdk::env::panic_str("The contract has already been initialized");
    }
    let contract = ContextConfigs::init();
    ::near_sdk::env::state_write(&contract);
}
```

so it does this already

File: contracts/proxy-lib/tests/config_helper.rs
Comment (danwilliams) [inline contracts/proxy-lib/tests/config_helper.rs]: More idiomatic to use the specific type names here, instead of `Default`.

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]:
```suggestion
        let signer_id = proposal.signer_id;
```

also, `SignerId: Copy`, so you can copy it here (i.e remove `&`) and reference everywhere else

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: `SignerId: Copy`, so:

```suggestion
        self.num_proposals_pk
            .insert(*singer_id, num_proposals);
```

or (if you removed `&` in the earlier comment)

```suggestion
        self.num_proposals_pk
            .insert(singer_id, num_proposals);
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: `create_proposal` will call `.parse` and so are you here.

you can introduce a non-`pub`* method that creates a proposal from a pre-parsed proposal.

*: I didn't say "private" since `near-sdk` has a concept of `#[private]` and that's not what I mean

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: do we want anybody to be able submit a proposal even if they're not a context member?

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]:
```suggestion
    fn internal_confirm(&mut self, request_id: ProposalId, signer_id: Repr<SignerId>) {
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: nit, shouldn't matter much:

```suggestion
        let mut requests = Vec::with_capacity((self.proposals.len() - offset).min(length));
        for request in self.proposals.iter().skip(offset).take(length) {
            requests.push(request);
        }
        requests
```

or leave it to `Iterator` which appropriately derive the exact size from `Iterator::size_hint`

```suggestion
        self.proposals.iter().skip(offset).take(length).collect()
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: shouldn't we return an option here if the proposal doesn't exist?

```suggestion
    pub fn get_confirmations_count(&self, proposal_id: ProposalId) -> Option<ProposalWithApprovals> {
        let size = self
            .approvals
            .get(&proposal_id)?
            .len();
```

File: contracts/proxy-lib/src/ext_config.rs
Comment (miraclx) [inline contracts/proxy-lib/src/ext_config.rs]: has this been implemented in the ctx config contract?

File: contracts/context-config/src/query.rs
Comment (xilosada) [reply contracts/context-config/src/query.rs:51]: I followed the structure of the other functions in the module. I had to add this helper to faciliate the cx contract call. cc: @miraclx

File: contracts/proxy-lib/src/ext_config.rs
Comment (xilosada) [reply contracts/proxy-lib/src/ext_config.rs]: Fixed

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: Will do it in a following commit, leaving this open till that

File: contracts/proxy-lib/tests/config_helper.rs
Comment (xilosada) [reply contracts/proxy-lib/tests/config_helper.rs]: I'm afraid this is coming from the config-contract, I'll create an issue to address it in a different commit. Leaving this open until the issue is created

File: contracts/context-config/src/query.rs
Comment (miraclx) [reply contracts/context-config/src/query.rs:51]: yeah, so NEAR contracts will only roll back a failed transaction if execution fails, and they fail only when the contract panics, unlike ours.

Returning a result will serialize the result as a success value, even if it's an `Err`, 🫠 

So the spent gas won't be refunded, state mutations will hold, etc..

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: +1 it must panic and keep the map empty. Will work on it

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: Fair

File: contracts/proxy-lib/src/ext_config.rs
Comment (xilosada) [reply contracts/proxy-lib/src/ext_config.rs]: here it goes https://github.com/calimero-network/core/pull/854/files#diff-22355c320575851613e81b59302954d94433115e28c09199b70079772934a535R47

Comment (miraclx) [reply contracts/proxy-lib/src/ext_config.rs]: Ah, missed that, noice.

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: it's `SignerId` on both sides, so no need to transmute

```suggestion
        let author_id = *proposal.author_id;
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]:
```suggestion
        let has_member = call_result.expect("Membership check failed");
        assert!(has_member, "Is not a member");
```

we typically don't need the `Error:` prefix, as that's already evident to whoever's handling it

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: Maybe we need to make this more dynamic?

```suggestion
            .with_static_gas(env::prepaid_gas())
```

or perhaps

```suggestion
            .with_static_gas(env::prepaid_gas().saturating_sub(env::used_gas()))
```

though I'd imagine the manual subtraction shouldn't be necessary since that delta would be different at the end of the execution anyway, so we may never be able to determine the exact amount of unused gas

does removing the `with_static_gas` definition fail?

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: it's cheaper to copy `SignerId` than to clone `Proposal`

```suggestion
        let author_id = *proposal.author_id;

        self.proposals.insert(proposal_id, proposal);
        self.approvals.insert(proposal_id, HashSet::new());

        self.internal_confirm(proposal_id, author_id);
```

File: contracts/proxy-lib/tests/common/proxy_lib_helper.rs
Comment (miraclx) [inline contracts/proxy-lib/tests/common/proxy_lib_helper.rs]: [`VerifyingKey: ReprBytes`](https://github.com/calimero-network/core/blob/fd4fdc9c6a8f52baf501ae5f013dbe698064b6e1/crates/context/config/src/types.rs#L255), so:

```suggestion
            author_id: author.verifying_key().rt().expect("Invalid signer"),
```

File: contracts/proxy-lib/tests/common/proxy_lib_helper.rs
Comment (miraclx) [inline contracts/proxy-lib/tests/common/proxy_lib_helper.rs]:
```suggestion
        let signed = Signed::new(&proposal, |p| author.sign(p))?;
```

File: contracts/proxy-lib/tests/common/proxy_lib_helper.rs
Comment (miraclx) [inline contracts/proxy-lib/tests/common/proxy_lib_helper.rs]: interesting model 😅 

typically we do something like:

```suggestion
            .await;
        call.map_err(Into::into)
```

but here you can just bubble it up at the point you got the result, so it's clearer

```suggestion
            .await?;
        Ok(call)
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs]: As discussed, you'll be introducing a mechanism for re-validating the approvals, yeah?

So `has_member` will need to become `has_members` so we can pass the list of approvers to it, yeah?

Or maybe pass a list of IDs, and return a list of non-members. So we can identify which approvals to invalidate.

Or instead of returning a list of non-members, it just returned a count of non-members. So knowing when to execute the request would be:

`if threshold == (number_of_members_checked - number_of_non_members_received) {}`

which also means if someone leaves the context and rejoins before the execution under the same identity, their approval stays valid

File: contracts/proxy-lib/src/ext_config.rs
Comment (miraclx) [inline contracts/proxy-lib/src/ext_config.rs:6]: also, btw, you can use this directly from the crate if you added it as a dependency

```rust
use calimero_context_config_near::ContextConfigs;

let promise = ContextConfigs::ext(self.context_config_account_id)
    .with_static_gas(..)
    .members(context_id, offset, length)
    .then(..);
```

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: What about this?

File: crates/context/config/src/repr.rs
Comment (xilosada) [reply crates/context/config/src/repr.rs]: derived Hash for Identity and SignerId

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: Done

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: Removed

File: contracts/proxy-lib/tests/common/proxy_lib_helper.rs
Comment (xilosada) [reply contracts/proxy-lib/tests/common/proxy_lib_helper.rs]: :D

File: contracts/proxy-lib/src/ext_config.rs
Comment (xilosada) [reply contracts/proxy-lib/src/ext_config.rs:6]: Will keep this until I change the config contract with the new apis

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs]: discussed internally. This will be part of a new PR

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [inline contracts/proxy-lib/src/lib.rs]: to IterableMap

File: contracts/proxy-lib/tests/common/proxy_lib_helper.rs
Comment (miraclx) [inline contracts/proxy-lib/tests/common/proxy_lib_helper.rs:46]:
```suggestion
        actions: &[ProposalAction],
```

File: contracts/proxy-lib/tests/common/counter_helper.rs
Comment (MatejVukosav) [inline contracts/proxy-lib/tests/common/counter_helper.rs]:
```suggestion
impl CounterContractHelper {
```

File: contracts/proxy-lib/tests/common/counter_helper.rs
Comment (MatejVukosav) [inline contracts/proxy-lib/tests/common/counter_helper.rs]:
```suggestion
pub struct CounterContractHelper {
```

File: contracts/proxy-lib/tests/common/mod.rs
Comment (MatejVukosav) [inline contracts/proxy-lib/tests/common/mod.rs:14]: Can this fail?

Comment (xilosada) [reply contracts/proxy-lib/tests/common/mod.rs:14]: yes and it will make the test fail

File: contracts/proxy-lib/tests/common/counter_helper.rs
Comment (xilosada) [reply contracts/proxy-lib/tests/common/counter_helper.rs]: good catch

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:15]: `ext_config` is already in scope

```suggestion
pub use ext_config::config_contract;
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:62]: you shouldn't need `Repr` in state at all

```suggestion
    pub signer_id: SignerId,
```

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:130]: Hm, they can take advantage of this only being processed in the next block to submit multiple proposals at once, hoping they'll be included in the same block, at which point they'll subvert this limitation

we should increment this number before the config check

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:236]: to match SDK expectations, proposal ID should be random and determined by the runtime, but this can come in a separate PR, just noting it

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:381]:
```suggestion
    ) -> Vec<(&[u8], &[u8])> {
```

File: crates/context/config/src/types.rs
Comment (miraclx) [inline crates/context/config/src/types.rs:1]: we shouldn't need these ..._

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:44]: _... if this was a `BTreeSet`

```suggestion
    pub approvals: IterableMap<ProposalId, BTreeSet<SignerId>>,
```

are you concerned about rebalancing on insertion since `SignerId` is random? 32-bytes is pretty small comparison

File: contracts/proxy-lib/src/lib.rs
Comment (miraclx) [inline contracts/proxy-lib/src/lib.rs:111]: should we take this as an initialization parameter?

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs:236]: yes

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs:111]: @petarjuki7 let's add it once we have the proxy-contract integration in the node

**Review by miraclx — APPROVED** [review]

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs:236]: @danwilliams plese change it once you are generating the Proposal Id

File: contracts/proxy-lib/src/lib.rs
Comment (xilosada) [reply contracts/proxy-lib/src/lib.rs:130]: will do later

**Pull request merged**
